Fix crash when 'run selected tests' but no tests are selected

rowo360 opened the pull request:
This PR fixes issue #1138.

Unfortunately I didn't manage to detect the actual root cause of this crash. As described in the issue repeating the same steps which were previously causing the crash suddently work. That doesn't feel good, but I'd rather fix the problem now than keep looking for a long time.

So the property `SelectedTests` might be set to null in some scenarios, which will cause the crash when executing 'Run selected tests'. Therefore I propose to fix it simply by handling this case now.

CharliePoole commented on src/TestCentric/testcentric.gui/Presenters/TestCentricPresenter.cs:
```diff
@@ -954,7 +954,8 @@ private void RunAllTests()
             var testSelection = _model.SelectedTests;
-            _model.RunTests(testSelection);
+            if (testSelection != null)
+                _model.RunTests(testSelection);
```
How are we able to get here if there are no selected tests? While I agree that the code should check for null or empty on general principle, the overall strategy in the UI has been to disable any controls that are not usable at a given point in time. So, if there are no selected tests,  this code ought to be unreachable.

CharliePoole reviewed (changes requested):
I can see that a test selection must  be capable of being empty but it doesn't seem to me that we should  be calling the model with an empty selection.

Additionally, if no tests are selected, it seems like the controls and menu items used to run selected tests should be disabled.

rowo360 commented:
That idea is even better: disable the toolbar item if no tests are selected! I like it!
I prepared a solution for this idea and added a new commit.
Basically I introduced a new event `SelectedTestsChanged` in the model. This event gets fired whenever some nodes are checked or selected in the tree and the TreeViewPresenter updates the `SelectedTests` property. The TestCentricPresenter class listens to this event and updates the enable state of the toolbar item accordingly.

Here are some screenshots to get a better impression:
In the first image the toolbar button is disabled because no node is checked, in the second image a node is checked and thus the toolbar button is enabled. The last image is without the option 'Show checkboxes' - in that case the toolbar button is always enabled - I didn't notice a scenario in which no node is selected.

<img src="https://github.com/user-attachments/assets/6f126b43-a23f-469c-8659-0f5a9791cf79" width="900">

rowo360 commented:
While working on this issue I wondered if the toolbar tooltip is always accurate?
The tooltip states 'Run selected tests' - but wouldn't it be more accurate to state 'Run checked tests' when the checkboxes are shown in the tree? If no checkboxes are shown, we can continue to use 'Run selected tests' as the tooltip. 
That might be helpful to understand why the toolbar button is disabled.

CharliePoole commented:
In my mind, when I first did this "selected" meant highlighted OR checked when checks are enabled. I agree with you that this might be confusing to the user. However, I think there are actually three states here, not two, in the way it is implemented.

1. If no checkboxes are shown, selected means selected, i.e. highlighted.
2. If checkboxes are shown _and nothing is checked_ it still means the same thing.
3. if checkboxes are shown and one or more tests are checked, then "selected" means checked.

I suggest you first verify that it works as I say, especially the second condition. If it does, then I think you would change the language in the tooltip each time it displays based on whether anything is checked rather than whether checkboxes are shown.

CharliePoole commented:
Is a new event needed? I'm pretty sure we were already responding to changed selection. It might be a generic property change to our settings.

CharliePoole commented on src/TestCentric/testcentric.gui/Presenters/TestCentricPresenter.cs:
```diff
@@ -754,10 +756,11 @@ private void UpdateViewCommands(bool testLoading = false)
             _view.RunParametersButton.Enabled = testLoaded && !testRunning;
 
+            _view.RunSelectedButton.Enabled = testLoaded && !testRunning && _model.SelectedTests != null && _model.SelectedTests.Any();
+
```
Will it let you use `_model.SelectedTests?.Any()` here>

rowo360 replied:
Unfortunately that's rejected by the c# compiler: 
Operator '&&' cannot be applied to operands of type 'bool' and 'bool?'

CharliePoole reviewed (approved):
Looks good to me. Thanks!

My one inline comment is at your option. If you want to change the tooltip, feel free to add another commit but check out my general comment about it first.

rowo360 commented:
> Is a new event needed? I'm pretty sure we were already responding to changed selection. It might be a generic property change to our settings.

Yes indeed, there's already a similar event: the model has a property ActiveTestItem and any change of this property triggers a SelectedItemChanged event. The ActiveTestItem is set by the TreeViewPresenter whenever the selected tree node changes. But checking/unchecking tree nodes will not change the tree node selection and thus not change the ActiveTestItem. But instead only the `SelectedTests` property of the model is updated. So, I added this missing event to the `SelectedTests` property to comply to the pattern.
BTW: there's also no setting change event while checking/unchecking tree nodes

CharliePoole commented:
FYI, once you merge the branch is pretty useless so you can delete it.
